src/coprocessor/statistics/analyze.rs

```diff
@@ -364,12 +377,62 @@ impl<S: Snapshot> RowSampleBuilder<S> {
         ))
     }
 
+    // Support tuning of cpu quota at present, io bandwidth tuning is on the way
+    // some smarter tuning algorithm like PID could be considered
+    fn tune_quota_limiter(&mut self, mut proc_stat: ProcessStat) {
+        debug_assert!(
+            self.quota_limiter.is_auto_tune_enabled(),
+            "auto_tune must be turned on!"
+        );
+        // rule based tuning:
+        //      1) if instance is busy, shrink cpu quota for analyze by 1 core until lower bound is hit;
+        //      2) if instance cpu usage is at healthy state, no op;
+        //      3) if instance is idle, increase cpu quota by 1 core until upper bound is hit.
+
+        let cpu_util = (proc_stat.cpu_usage().unwrap() / 100.0) / SysQuota::cpu_cores_quota();
+
+        if cpu_util >= SYSTEM_BUSY_THRESHOLD {
+            self.quota_limiter.set_cpu_time_limit(
+                1_000_000
+                    * cmp::max(
+                        (self.quota_limiter.cputime_limiter() / 1_000_000_000.0) as usize - 1,
+                        ANALYZER_CORE_LOWER_BOUND,
+                    ),
+            );
+            debug!("cpu_time_limit shrinked";
+                            "cpu_util" => ?cpu_util,
+                            "new quota" => ?self.quota_limiter.cputime_limiter() / 1_000_000_000.0);
+            NON_TXN_COMMAND_THROTTLE_TIME_COUNTER_VEC_STATIC
```

Comment:
Needs a metrics for the limit. Previously we don't need one because the limit is fixed.

Reply:
I don't see the value of adding metrics for this session level quota change.